File: python/revo3/streaming_control.py
"""Open a Motion streaming-control session and hold the current joint positions."""

import argparse
import asyncio
import time

from bc_revo3_sdk import main_mod as sdk


async def run(args: argparse.Namespace) -> None:
    if not args.run:
        raise RuntimeError("Pass --run after checking the work area and independent stop path")

    manager = sdk.Manager()
    hand = None
    session = None
    try:
        hand = await manager.connect_auto(port=args.port, slave_id=args.slave_id)
        layout = hand.joint_layout
        if layout is None or layout.joint_count != 21:
            raise RuntimeError("The current streaming-control preview requires a 21-joint layout")

        state = await hand.state.snapshot()
        hold_positions = list(state.positions_deg)
        session = hand.motion.open_servo(command_timeout_ms=100)

        period = 1.0 / args.rate
        deadline = time.monotonic() + args.duration
        while time.monotonic() < deadline:
            await session.send_position(hold_positions)
            await asyncio.sleep(period)

        print(f"Servo session state before close: {session.state}")
    finally:
        if session is not None:
            session.close()
        if hand is not None:
            await hand.close()
        await manager.close()


def parse_args() -> argparse.Namespace:
    parser = argparse.ArgumentParser()
    parser.add_argument("--port")
    parser.add_argument("--slave-id", type=lambda value: int(value, 0))
    parser.add_argument("--duration", type=float, default=1.0)
    parser.add_argument("--rate", type=float, default=50.0)
    parser.add_argument("--run", action="store_true")
    return parser.parse_args()


if __name__ == "__main__":
    asyncio.run(run(parse_args()))

File: c/demo/streaming_control.cpp
#include <revo3/revo3.hpp>

#include <atomic>
#include <chrono>
#include <csignal>
#include <cstdio>
#include <cstring>
#include <stdexcept>
#include <thread>
#include <vector>

namespace {
std::atomic<bool> g_running{true};
void signal_handler(int) { g_running = false; }
}

int main(int argc, char **argv) {
  revo3::init_logging(LOG_LEVEL_INFO, true);
  std::signal(SIGINT, signal_handler);
  using namespace std::chrono_literals;

  bool run = false;
  revo3::DiscoveryOptions discovery;
  for (int index = 1; index < argc; ++index) {
    if (std::strcmp(argv[index], "--move") == 0 || std::strcmp(argv[index], "--run") == 0) {
      run = true;
    } else {
      discovery.port = argv[index];
    }
  }
  if (!run) {
    std::fprintf(stderr,
                 "Pass --move after checking the work area and independent "
                 "stop path\n");
    return 2;
  }

  try {
    revo3::Manager manager;
    auto hand = manager.connect_auto(discovery);
    const auto layout = hand.joint_layout();
    if (!layout) {
      throw std::runtime_error("Joint layout is unavailable");
    }
    const auto state = hand.state().snapshot();
    std::vector<float> hold(state.motors.positions_deg,
                            state.motors.positions_deg + layout->joint_count);

    auto session = hand.motion().open_servo();
    const auto deadline = std::chrono::steady_clock::now() + 1s;
    // Break loop gracefully on SIGINT (Ctrl+C)
    while (g_running && std::chrono::steady_clock::now() < deadline) {
      session.send_position(hold);
      std::this_thread::sleep_for(20ms);
    }
    std::printf("Servo session state before close: %d\n",
                static_cast<int>(session.state()));
    session.close();
    return 0;
  } catch (const std::exception &error) {
    std::fprintf(stderr, "Error: %s\n", error.what());
    return 1;
  }
}
